### PR TITLE
Optimize TimestampMapping::leader to avoid stream overhead

### PR DESCRIPTION
**Pull request by schlosna**
## General
**Before this PR**:

`com.palantir.atlasdb.keyvalue.api.watch.TimestampMapping.leader():50` is on a hot path finding the distinct leader for given `LockWatchVersion`s and could be optimized

image

**After this PR**:

==COMMIT_MSG==
Optimize TimestampMapping::leader to avoid stream overhead
==COMMIT_MSG==

**Priority**:

**Concerns / possible downsides (what feedback would you like?)**:

**Is documentation needed?**:

## Compatibility
**Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?**:

**Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?**:

**The code in this PR may be part of a blue-green deploy. Can upgrades from previous versions safely coexist? (Consider restarts of blue or green nodes.)**:

**Does this PR rely on statements being true about other products at a deployment - if so, do we have correct product dependencies on these products (or other ways of verifying that these statements are true)?**:

**Does this PR need a schema migration?**

## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**:

**What was existing testing like? What have you done to improve it?**:

**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**:

**If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?**:

## Execution
**How would I tell this PR works in production? (Metrics, logs, etc.)**:

**Has the safety of all log arguments been decided correctly?**:

**Will this change significantly affect our spending on metrics or logs?**:

**How would I tell that this PR does not work in production? (monitors, etc.)**:

**If this PR does not work as expected, how do I fix that state? Would rollback be straightforward?**:

**If the above plan is more complex than “recall and rollback”, please tag the support PoC here (if it is the end of the week, tag both the current and next PoC)**:

## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**:

**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**:

**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**:

## Development Process
**Where should we start reviewing?**:

**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**:

**Please tag any other people who should be aware of this PR**:
@jeremyk-91
@sverma30
@raiju

### CODE REVIEW

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/TimestampMapping.java [hunk 1]
```diff
@@ -31,7 +35,7 @@ interface TimestampMapping {
 
     @Value.Derived
     default Range<Long> versionRange() {
-        LongSummaryStatistics summary = timestampMapping().values().stream()
+        LongSummaryStatistics summary = checkNotEmpty(timestampMapping()).values().stream()
```
Comment (mdaudali) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/TimestampMapping.java:38]: I believe `checkNotEmpty` shouldn't be necessary given the `nonEmptyMapping` check?

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/TimestampMapping.java [hunk 2]
```diff
@@ -44,18 +48,36 @@ default LockWatchVersion lastVersion() {
...
 
     static ImmutableTimestampMapping.Builder builder() {
         return ImmutableTimestampMapping.builder();
     }
+
+    @VisibleForTesting
+    static <T> T distinctElement(Iterable<? extends T> iterable) {
+        Iterator<? extends T> iterator = iterable.iterator();
+        if (!iterator.hasNext()) {
+            throw new SafeIllegalStateException("No elements");
```
Comment (mdaudali) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/TimestampMapping.java:73]: Noting for posterity, but I don't actually think it matters - we're changing from MoreCollectors NoSuchElementException and IllegalArgumentException to IllegalStateException

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/TimestampMappingTest.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+import com.palantir.lock.watch.LockWatchVersion;
+import com.palantir.logsafe.exceptions.SafeIllegalArgumentException;
+import com.palantir.logsafe.exceptions.SafeIllegalStateException;
+import java.util.UUID;
+import org.junit.Test;
+
+public class TimestampMappingTest {
+    @Test
+    public void single_leader_mapping_is_valid() {
```
Comment (mdaudali) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/TimestampMappingTest.java:30]: non-blocking, but we generally stick with camel case naming for our tests :)

(Wonder if this is something we could enforce via a lint...)

**Review by mdaudali — COMMENTED** [review]
Left a couple comments, although I appreciate this is a draft! Thanks for the contribution :)

Please ping when you're done and want an approval!

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/TimestampMapping.java (hunk 1 above)
Comment (schlosna) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/TimestampMapping.java:38]: Good call, will remove

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/TimestampMappingTest.java (hunk 3 above)
Comment (schlosna) [reply atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/TimestampMappingTest.java:30]: I can update.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/TimestampMapping.java (hunk 2 above)
Comment (schlosna) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/TimestampMapping.java]: updated to maintain exception back compat (though throwing SafeIllegalArgumentException with mismatched IDs)

**Comment by schlosna** [comment]
thanks for the review @mdaudali , updated to reflect your comments

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/TimestampMappingTest.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+    @Test
+    public void emptyDistinctElements() {
+        assertThatThrownBy(() -> TimestampMapping.identicalElement(List.of()))
+                .isInstanceOf(NoSuchElementException.class);
+    }
+
+    @Test
+    public void multipleDistinctElements() {
+        assertThatThrownBy(() -> TimestampMapping.identicalElement(List.of("a", "b", "a")))
```
Comment (mdaudali) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/TimestampMappingTest.java:85]: nit: if assertThatLoggableException is available, lets use that? We can then also ensure the args are correct.

**Review by mdaudali — COMMENTED** [review]
lgtm, thanks!

**Review by mdaudali — APPROVED** [review]

**Pull request merged**